rosidl_typesupport_c/resource/idl__type_support.cpp.em

```diff
@@ -0,0 +1,100 @@
+// with input from @(package_name):@(interface_path)
+// generated code does not contain a copyright notice
+@
+@#######################################################################
+@# EmPy template for generating <idl>__type_support.cpp files
+@#
+@# Context:
+@#  - package_name (string)
+@#  - interface_path (Path relative to the directory named after the package)
+@#  - interfaces (list of interfaces, either Messages or Services)
+@#  - type_supports (list of strings, the names of the type support packages)
+@#######################################################################
+@{
+include_directives = set()
+}@
+@#######################################################################
+@# Handle message
+@#######################################################################
+@{
+from rosidl_parser.definition import Message
+}@
+@[for message in content.get_elements_of_type(Message)]@
+
+@{
+TEMPLATE(
+    'msg__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path, message=message,
+    include_directives=include_directives, type_supports=type_supports)
+}@
+@[end for]@
+@
+@#######################################################################
+@# Handle service
+@#######################################################################
+@{
+from rosidl_parser.definition import Service
+}@
+@[for service in content.get_elements_of_type(Service)]@
+
+@{
+TEMPLATE(
+    'msg__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path,
+    message=service.request_message, include_directives=include_directives,
+    type_supports=type_supports)
+}@
+
+@{
+TEMPLATE(
+    'msg__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path,
+    message=service.response_message, include_directives=include_directives,
+    type_supports=type_supports)
+}@
+
+@{
+TEMPLATE(
+    'srv__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path, service=service,
+    include_directives=include_directives, type_supports=type_supports)
+}@
+@[end for]@
+@
+@#######################################################################
+@# Handle action
+@#######################################################################
+@{
+from rosidl_parser.definition import Action
+}@
+@[for action in content.get_elements_of_type(Action)]@
+
+@{
+TEMPLATE(
+    'srv__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path, service=action.goal_service,
+    include_directives=include_directives, type_supports=type_supports)
+}@
+
+@{
+TEMPLATE(
+    'srv__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path, service=action.result_service,
+    include_directives=include_directives, type_supports=type_supports)
+}@
+
+@{
+TEMPLATE(
+    'msg__type_support.cpp.em',
+    package_name=package_name, interface_path=interface_path, message=action.feedback,
```

Comment:
IIUC `action.feedback_message` wraps `action.feedback`. In that case, I *think* the actual topic type is `action.feedback_message` and thus the only type `rmw` implementations deal with.

Reply:
That will require to change the action struct in https://github.com/ros2/rosidl/pull/334/files#diff-45bcd58679cbbef4edcbf7b568fceb46R31 which doesn't contain the FeedbackMessage atm. Without any code using these pieces this becomes rather fuzzy. I think we need to first establish what the action struct is used for and which types it needs to provide.

Reply:
> That depends which one the higher level action code will need?

I think @hidmic is correct that the generated template needs the wrapper `action.feedback_message` since that's what's used to create the publisher and subscriber. The question is what generates the struct for the message with the user's fields? It seems like this template needs to invoke the message template twice: once for the message with the user's fields and once for the wrapper message.

> I think we need to first establish what the action struct is used for and which types it needs to provide.

The struct for an action server will need to contain both: The wrapper message type needs to be known by `rclcpp_action` so it can grab the user's nested field, and the user's type needs to be in the struct so it has a type to pass to a user's callback.

Reply:
Since you know how you plan to use this can you propose a specific API?

Reply:
I extended the struct in https://github.com/ros2/rosidl/pull/334/commits/1dcf9127866d2af37285be1c752139280368c7b4 and updated this PR to use the new `FeedbackMessage` type: f19110f54153d3daaa1a2cafc29776a3ce444249